### DISCUSSION TITLE
Expert Opinions Please - Data Centric (Generic) API

### DISCUSSION THREAD

**Discussion by JumpinJackFlash** [OAI/OpenAPI-Specification#2990] [General]
Hello,

I'd like to elicit some opinions from the experts on how I'm approaching API design and how OAI might be applicable.
First off, I have a framework and middle-tier components that allow me to push all user resource (code & data, including unstructured) into a database (Oracle...get your arrows and slings out..!!!). There's nothing in the middle-tier in my design other than a protocol adapter between HTTP and SQL.

So...what that means is I can build an API, expressed as a package in the database, that only exposes a single entry point. Remember, all of the logic is in the DB. That means all of your routing is there too. That's what allows you to expose a single entry point. The logic in the database looks at your 'message' and figures out what to do.

This is what I call a Data Centric (Generic) API.

Here's my package header which is essentially the API signature:

`create or replace package db_twig as

  function call_rest_api
  (
    p_json_parameters                 clob
  )
  return clob;

end db_twig;
`
As you can see, not much to expose.
So, my questions are:

- How would OAI deal with API components that are embedded within p_json_parameters, which is a JSON string sent into the function in the DB?
- It seems as though OAI is geared towards APIs that expose their components as entry-points and parameters. Would it even be possible to use OAI to test APIs developed in the manner above?

Thanks for your time and opinions on this matter.

**Comment by handrews** [comment]
As you note, this doesn't seem to fit the current OAS paradigm very well.  However, you may find the [Moonwalk (OAS4) discussions](https://github.com/OAI/sig-moonwalk/discussions) around [signatures](https://www.openapis.org/blog/2023/12/06/openapi-moonwalk-2024) to be of interest.
